pkg/tracker/reward/querier.go

```diff
@@ -0,0 +1,170 @@
+)
+
+type RewardQuerier struct {
+	client           *ethclient.Client
+	logger           log.Logger
+	contractInstance *contracts.ITellor
+	ctx              context.Context
+	stop             context.CancelFunc
+
+	tsDB   storage.SampleAndChunkQueryable
+	aggr   aggregator.IAggregator
+	engine *promql.Engine
+}
+
+func NewRewardQuerier(
+	logger log.Logger,
+	ctx context.Context,
+	cfg Config,
+	tsDB storage.SampleAndChunkQueryable,
+	client *ethclient.Client,
+	contractInstance *contracts.ITellor,
+	addr common.Address,
+	aggr aggregator.IAggregator,
+) (*RewardQuerier, error) {
+	logger, err := logging.ApplyFilter(cfg.LogLevel, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "apply filter logger")
+	}
+	logger = log.With(logger, "component", ComponentName)
+
+	opts := promql.EngineOpts{
+		Logger:               logger,
+		Reg:                  nil,
+		MaxSamples:           30000,
+		Timeout:              10 * time.Second,
+		LookbackDelta:        5 * time.Minute,
+		EnableAtModifier:     true,
+		EnableNegativeOffset: true,
+	}
+	engine := promql.NewEngine(opts)
+
+	ctx, cncl := context.WithCancel(ctx)
+	return &RewardQuerier{
+		client:           client,
+		logger:           logger,
+		contractInstance: contractInstance,
+		ctx:              ctx,
+		stop:             cncl,
+		tsDB:             tsDB,
+		engine:           engine,
+		aggr:             aggr,
+	}, nil
+}
+
+// Current returns the profit in percents based on the current TRB price.
+func (self *RewardQuerier) Current(ctx context.Context, slot *big.Int, gasPriceEth1e18 *big.Int) (int64, error) {
+	gasUsed, err := self.GasUsed(ctx, slot)
+	if err != nil {
+		return 0, err
+	}
+
+	rewardEth1e18, err := self.rewardInEth1e18()
+	if err != nil {
+		return 0, errors.New("getting trb current TRB price")
+	}
+
+	txCostEth1e18 := big.NewInt(0).Mul(gasPriceEth1e18, gasUsed)
+	profit := big.NewInt(0).Sub(rewardEth1e18, txCostEth1e18)
+	profitPercentFloat := float64(profit.Int64()) / float64(txCostEth1e18.Int64()) * 100
+	profitPercent := int64(profitPercentFloat)
+
+	level.Debug(self.logger).Log(
+		"msg", "profit checking",
+		"reward", fmt.Sprintf("%.2e", float64(rewardEth1e18.Int64())),
+		"txCost", fmt.Sprintf("%.2e", float64(txCostEth1e18.Int64())),
+		"slot", slot,
+		"gasUsed", gasUsed,
+		"gasPrice", gasPriceEth1e18,
+		"profit", fmt.Sprintf("%.2e", float64(profit.Int64())),
+		"profitMargin", profitPercent,
+	)
+
+	return profitPercent, nil
+}
+
+type ErrNoDataForSlot struct {
+	slot string
+}
+
+func (e ErrNoDataForSlot) Error() string {
+	return "no data for gas used for slot:" + e.slot
+}
+
+// GasUsed estimates the gas needed by the transaction.
+func (self *RewardQuerier) GasUsed(ctx context.Context, slot *big.Int) (*big.Int, error) {
+	query, err := self.engine.NewInstantQuery(
+		self.tsDB,
+		`last_over_time(gas_usage_estimated{slot="`+slot.String()+`"}[1d]) - `+`last_over_time(gas_usage_actual{slot="`+slot.String()+`"}[1d])`,
```

Comment:
Why is this like this?
This query will return the refund no? we need gas_usage_estimated - the refund right?